src/lib/annotations/drawing/DrawingController.js

```diff
@@ -0,0 +1,303 @@
+                type: 'click',
+                func: () => {
+                    this.currentThread.saveAnnotation(TYPES.draw);
+                    this.annotator.toggleAnnotationHandler(TYPES.draw);
+                },
+                eventObj: this.postButtonEl
+            });
+        }
+
+        if (this.undoButtonEl) {
+            handlers.push({
+                type: 'click',
+                func: () => {
+                    this.currentThread.undo();
+                },
+                eventObj: this.undoButtonEl
+            });
+        }
+
+        if (this.redoButtonEl) {
+            handlers.push({
+                type: 'click',
+                func: () => {
+                    this.currentThread.redo();
+                },
+                eventObj: this.redoButtonEl
+            });
+        }
+
+        return handlers;
+    }
+
+    /**
+     * Handle an annotation event.
+     *
+     * @inheritdoc
+     * @protected
+     * @param {AnnotationThread} thread - The thread that emitted the event
+     * @param {Object} data - Extra data related to the annotation event
+     * @return {void}
+     */
+    handleAnnotationEvent(thread, data = {}) {
+        switch (data.type) {
+            case 'locationassigned':
+                // Register the thread to the threadmap when a starting location is assigned. Should only occur once.
+                this.annotator.addThreadToMap(thread);
+                break;
+            case 'drawcommit':
+                // Upon a commit, remove the listeners on the thread.
+                // Adding the thread to the Rbush only happens upon a successful save
+                thread.removeAllListeners('annotationevent');
+                break;
+            case 'pagechanged':
+                // On page change, save the original thread, create a new thread and
+                // start drawing at the location indicating the page change
+                this.currentThread = undefined;
+                thread.saveAnnotation(TYPES.draw);
+                this.unbindModeListeners();
+                this.bindModeListeners(TYPES.draw);
+                this.currentThread.handleStart(data.location);
+                break;
+            case 'availableactions':
+                this.updateUndoRedoButtonEls(data.undo, data.redo);
+                break;
+            default:
+        }
+    }
+
+    /**
+     * Find the selected drawing threads given a pointer event. Randomly picks one if multiple drawings overlap
+     *
+     * @protected
+     * @param {Event} event - The event object containing the pointer information
+     * @return {void}
+     */
+    handleSelection(event) {
+        if (!event) {
+            return;
+        }
+
+        const location = this.annotator.getLocationFromEvent(event, TYPES.point);
+        if (!location) {
+            return;
+        }
+
+        const eventBoundary = {
+            minX: +location.x - DRAW_BORDER_OFFSET,
```

Comment:
What do these +'s do?

Reply:
Sometimes location.x/y are strings (I suspect this is when they are loaded from the server) and + ensures they are processed as numbers. We can also use parseFloat() here if that is more clear.

Reply:
It looks like +number is actually the fastest. https://jsperf.com/number-vs-plus-vs-toint-vs-tofloat/16